File: crates/spoke-connect/src/core/nonce.rs
//! Single-use `(peer_id, nonce)` replay store.
//!
//! Normative rules (`.mstar/specs/spoke-connect.md` §Nonce / replay
//! protection): nonce uniqueness is scoped per **sender** `peer_id`; a
//! receiver MUST reject a hello whose `(peer_id, nonce)` pair was already
//! accepted. An in-memory set for the life of the process is sufficient for
//! protocol v1 (products MAY persist nonces with TTL).
//!
//! The store only records **accepted** hellos — the caller must not call
//! `check_and_record` for a hello that failed an earlier gate, so a rejected
//! hello stays retry-safe.

use std::collections::HashSet;

/// In-memory `(peer_id, nonce)` store of accepted hellos.
#[derive(Debug, Default)]
pub struct NonceStore {
    seen: HashSet<(String, String)>,
}

impl NonceStore {
    /// Creates an empty store.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Records `(peer_id, nonce)` unless it was already accepted; returns
    /// `false` on replay. Call only after the hello passed every earlier
    /// gate (allowlist, signature) so a rejected hello is not burned.
    pub fn check_and_record(&mut self, peer_id: &str, nonce: &str) -> bool {
        self.seen.insert((peer_id.to_owned(), nonce.to_owned()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn nonce_is_single_use_per_peer() {
        let mut store = NonceStore::new();
        assert!(store.check_and_record("peer-a", "nonce-1"));
        assert!(
            !store.check_and_record("peer-a", "nonce-1"),
            "replay rejected"
        );
        // A different nonce from the same peer is fresh.
        assert!(store.check_and_record("peer-a", "nonce-2"));
    }

    #[test]
    fn same_nonce_from_different_peers_is_allowed() {
        let mut store = NonceStore::new();
        assert!(store.check_and_record("peer-a", "shared-nonce"));
        assert!(
            store.check_and_record("peer-b", "shared-nonce"),
            "nonce scoping is per sender peer_id"
        );
    }

    #[test]
    fn fresh_store_accepts_everything() {
        let mut store = NonceStore::new();
        assert!(store.check_and_record("peer-a", "n-1"));
        assert!(store.check_and_record("peer-b", "n-1"));
        assert!(store.check_and_record("peer-b", "n-2"));
    }
}

File: packages/spoke-connect-ts/src/core/nonce.ts
/**
 * Single-use `(peer_id, nonce)` replay store + hello nonce generator.
 *
 * Ported from `crates/spoke-connect/src/core/nonce.rs`; normative rules
 * `.mstar/specs/spoke-connect.md` §Nonce / replay protection: nonce
 * uniqueness is scoped per **sender** `peer_id`; a receiver MUST reject a
 * hello whose `(peer_id, nonce)` pair was already accepted. An in-memory set
 * for the life of the process is sufficient for protocol v1.
 *
 * The store only records **accepted** hellos — the caller must not call
 * `checkAndRecord` for a hello that failed an earlier gate, so a rejected
 * hello stays retry-safe.
 */

import { base64UrlEncode } from "../crypto.js";

/**
 * In-memory `(peer_id, nonce)` store of accepted hellos. The composite key
 * uses a NUL separator — peer ids (base58btc) and nonces (base64url/hex)
 * never contain NUL, so the pair is unambiguous.
 */
export class NonceStore {
  private readonly seen = new Set<string>();

  /**
   * Records `(peerId, nonce)` unless it was already accepted; returns
   * `false` on replay. Call only after the hello passed every earlier gate
   * (allowlist, signature) so a rejected hello is not burned.
   */
  checkAndRecord(peerId: string, nonce: string): boolean {
    const key = `${peerId}\u0000${nonce}`;
    if (this.seen.has(key)) {
      return false;
    }
    this.seen.add(key);
    return true;
  }
}

/**
 * Fresh hello nonce: 16 CSPRNG bytes encoded base64url without padding
 * (22 chars, ≥ the minLength 16 wire floor; ≥128 bits entropy as the spec
 * recommends for generators).
 */
export function generateNonce(): string {
  const bytes = new Uint8Array(16);
  globalThis.crypto.getRandomValues(bytes);
  return base64UrlEncode(bytes);
}
